experimental/trusted_shuffler/trusted_shuffler/src/lib.rs

```diff
@@ -16,4 +16,131 @@
+use futures::future::join_all;
+use rand::{seq::SliceRandom, thread_rng};
+use std::{
+    future::Future,
+    marker::{Send, Sync},
+    mem::replace,
+    ops::DerefMut,
+    sync::{Arc, Mutex},
+};
+use tokio::sync::oneshot;
+
+type Request = Vec<u8>;
+type Response = Vec<u8>;
+
+struct Message {
+    // Arbitrary message data.
+    data: Vec<u8>,
+    // Channel for sending responses back to the client async tasks.
+    response_sender: oneshot::Sender<Response>,
+}
+
+// Trusted Shuffler inmpelentation.
+pub struct TrustedShuffler<F, O>
+where
+    F: Send + Sync + 'static + FnOnce(Request) -> O,
+    O: Send + Sync + Future<Output = Response>,
+{
+    // Value k that represents k-anonymity.
+    anonymity_value: usize,
+
+    // Current batch of requests to be shuffled.
+    current_requests: Arc<Mutex<Vec<Message>>>,
+
+    // Async function that sends a request to the backend and returns a response.
+    request_sender: F,
+}
+
+impl<F, O> TrustedShuffler<F, O>
+where
+    F: Send + Sync + Clone + 'static + FnOnce(Request) -> O,
+    O: Send + Sync + Future<Output = Response>,
+{
+    pub fn new(anonymity_value: usize, request_sender: F) -> Self {
+        Self {
+            anonymity_value,
+            current_requests: Arc::new(Mutex::new(vec![])),
+            request_sender,
+        }
+    }
+
+    // Asynchronously handles an incoming request.
+    pub async fn invoke(&self, request: Request) -> anyhow::Result<Response> {
+        let (sender, receiver) = oneshot::channel();
+
+        let request = Message {
+            data: request,
+            response_sender: sender,
+        };
+
+        // Check if the request batch is filled and create a separate task for shuffling requests.
+        {
+            let mut current_requests = self
+                .current_requests
+                .lock()
+                .map_err(|error| anyhow!("Couldn't lock current messages mutex: {:?}", error))?;
+            current_requests.push(request);
+
+            if current_requests.len() >= self.anonymity_value {
+                // Replace current requests with an empty vector.
+                let requests = replace(current_requests.deref_mut(), vec![]);
+
+                // Shuffle current requests and send them to the backend.
+                let request_sender = self.request_sender.clone();
+                tokio::spawn(async move {
+                    if let Err(error) =
+                        TrustedShuffler::shuffle_requests(requests, request_sender).await
+                    {
+                        log::error!("Shuffling error: {:?}", error);
+                    }
+                });
+            }
+        }
+
+        receiver.await.context("Couldn't receive response")
+    }
+
+    // Shuffles requests and sends them to the backend using the
+    // [`TrustedShuffler::request_sender`].
+    async fn shuffle_requests(mut requests: Vec<Message>, request_sender: F) -> anyhow::Result<()> {
```

Comment:
It seems a bit odd that this fn is defined on the struct itself, and, because it does not actually take self as arg, it needs the caller to manually pass the request_sender param which is actually on self. At that point, may as well make this a stand-alone function. I guess this saves some repetition of the trait bounds, but that's even less of an issue if using the trait as I suggested above. WDYT?

Reply:
It would create new duplication, because in the new function I'll need to also add
```rust
where
    F: Send + Sync + Clone + 'static + FnOnce(Request) -> O,
    O: Send + Sync + Future<Output = Response>,
```